### PR TITLE
Add Stockholm public transport extension

### PR DESCRIPTION
**Pull request by albingroen**
## Description

An extension where you can view public transport departures in Stockholm

- Search for stations
- Save stations as favorites
- Search departures for a station
- Filter on transport method (Metro, Train, Tram, Bus, Ship)

## Screencast

https://user-images.githubusercontent.com/19674362/232210480-4b3744cd-d8d0-4905-8584-c99604cebbf0.mp4

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

### CODE REVIEW

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi 👋 

Thanks for your contribution 💪 

I have now tested your extension, and I have some feedback ready for you:

* Seems like you need three letters to search, maybe it's better to first search when there are three letters?

* You could add a simple [List.EmptyView](https://developers.raycast.com/api-reference/user-interface/list#list.emptyview) with the same logo in 128x128px - that would spice it a bit up 😊

* In this view I can't search for Ropsten, maybe you need to add them to the keywords array on the `List.Item`

Raycast-FDxHqSVG@2x

* Some items are duplicated
Raycast-0e3kwn9i@2x

```
09:18:21.890 Warning: Encountered two children with the same key, `Petermatsvägen-Norrtälje`. Keys should be unique so that components maintain their identity across updates. Non-unique keys may cause children to be duplicated and/or omitted — the behavior is unsupported and could change in a future version.
```

I'm looking forward to testing this extension again 🔥 

Request a new review when you are ready. Feel free to contact me here or at [Slack](https://www.raycast.com/community) if you have any questions.

**Comment by albingroen** [comment]
Thanks for the review, will look it over later today!

File: extensions/stockholm-public-transport/src/components/stations.tsx [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+              station={station}
+            />
+          ))}
+        </List.Section>
+      )}
+
+      <List.EmptyView icon={Icon.Geopin} title="No stations found" />
+    </List>
+  );
+}
```
Comment (pernielsentikaer) [inline extensions/stockholm-public-transport/src/components/stations.tsx:82]:
```suggestion
import Station from "./station";
import useSWR from "swr";
import { IStation } from "../types";
import { List, Toast, showToast, Icon } from "@raycast/api";
import { ReactNode, useState } from "react";
import { addStationToFavorites, getFavoritStations, getStations, removeStationFromFavorites } from "../lib/stations";

interface StationsProps {
  onSelectStation: (station: IStation) => ReactNode;
}

export default function Stations({ onSelectStation }: StationsProps) {
  // Local state
  const [search, setSearch] = useState<string>("");

  // Server state
  const {
    data: stationsData,
    isLoading: isStationsLoading,
    isValidating: isStationsValidating,
  } = useSWR(["stations", search], () => getStations(search), {
    keepPreviousData: true,
    onError: (error) => {
      showToast({
        title: error.response?.data?.title || "Failed to retrieve stations",
        message: error.response?.data?.description,
        style: Toast.Style.Failure,
      });
    },
  });

  const {
    data: favoriteStationsData,
    isLoading: isFavoriteStationsLoading,
    isValidating: isFavoriteStationsValidating,
  } = useSWR("favorite-stations", getFavoritStations);

  const isLoading =
    isStationsLoading || isStationsValidating || isFavoriteStationsLoading || isFavoriteStationsValidating;

  return (
    <List
      filtering={false}
      isLoading={isLoading}
      onSearchTextChange={setSearch}
      searchBarPlaceholder="Search stations"
      throttle
    >
      {stationsData && (
        <List.Section title="All stations">
          {stationsData.map((station) => {
            const isFavorite = favoriteStationsData?.some(({ SiteId }) => SiteId === station.SiteId);

            return (
              <Station
                key={[station.Name, station.Location, station.SiteId].join("-")}
                onSelect={onSelectStation}
                onToggleFavorite={isFavorite ? removeStationFromFavorites : addStationToFavorites}
                isFavorite={isFavorite}
                station={station}
              />
            );
          })}
        </List.Section>
      )}

      {favoriteStationsData && (
        <List.Section title="Favorite stations">
          {favoriteStationsData.map((station) => (
            <Station
              isFavorite
              key={["favorites", station.Name, station.Location, station.SiteId].join("-")}
              onSelect={onSelectStation}
              onToggleFavorite={removeStationFromFavorites}
              station={station}
            />
          ))}
        </List.Section>
      )}

      <List.EmptyView
        icon={Icon.Geopin}
        title={isLoading && search.length >= 3 ? "Loading..." : "No Stations Found"}
        description={isLoading && search.length >= 3 ? "Please wait" : "Type at least three charecters to search"}
      />
    </List>
  );
}
```

File: extensions/stockholm-public-transport/src/lib/stations.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+      mutate("favorite-stations");
+    }
+  } catch {
+    showToast({
+      title: "Failed to remove from favorite stations",
+      style: Toast.Style.Failure,
+    });
+  }
+}
```
Comment (pernielsentikaer) [inline extensions/stockholm-public-transport/src/lib/stations.ts:84]:
```suggestion
import axios from "axios";
import { API_URL } from "./constants";
import { IStation } from "../types";
import { LocalStorage, showToast, Toast } from "@raycast/api";
import { mutate } from "swr";

export async function getStations(search: string) {
  if (!search || typeof search !== "string" || search.length < 3) {
    return [];
  }

  return axios.get<IStation[]>(API_URL + `/stations?search=${search}`).then((res) => res.data);
}

export async function getFavoritStations(): Promise<IStation[]> {
  try {
    const _stations = await LocalStorage.getItem<string>("favorite-stations");

    if (_stations) {
      return JSON.parse(_stations);
    }
  } catch {
    console.info("No favorite stations");
  }

  return [];
}

export async function addStationToFavorites(station: IStation) {
  let favorites: IStation[] = [];

  try {
    const _favorites = await LocalStorage.getItem<string>("favorite-stations");

    if (_favorites) {
      favorites = JSON.parse(_favorites);
    }
  } catch {
    console.info("No favorite stations");
  }

  try {
    await LocalStorage.setItem("favorite-stations", JSON.stringify([...favorites, station]));

    showToast({
      title: `Added ${station.Name} to favorite stations`,
      style: Toast.Style.Success,
    });

    mutate("favorite-stations");
  } catch {
    showToast({
      title: "Failed to add to favorite stations",
      style: Toast.Style.Failure,
    });
  }
}

export async function removeStationFromFavorites(station: IStation) {
  try {
    const _favorites = await LocalStorage.getItem<string>("favorite-stations");

    if (_favorites) {
      const favorites: IStation[] = JSON.parse(_favorites);

      await LocalStorage.setItem(
        "favorite-stations",
        JSON.stringify(favorites.filter(({ SiteId }) => SiteId !== station.SiteId))
      );

      showToast({
        title: `Removed ${station.Name} from favorite stations`,
        style: Toast.Style.Success,
      });

      mutate("favorite-stations");
    }
  } catch {
    showToast({
      title: "Failed to remove from favorite stations",
      style: Toast.Style.Failure,
    });
  }
}
```

File: extensions/stockholm-public-transport/src/components/station.tsx [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+              onToggleFavorite(station);
+            }}
+          />
+        </ActionPanel>
+      }
+      subtitle={station.Location}
+      title={station.Name}
+    />
+  );
+}
```
Comment (pernielsentikaer) [inline extensions/stockholm-public-transport/src/components/station.tsx:40]:
```suggestion
import { Action, ActionPanel, Color, Icon, List } from "@raycast/api";
import { IStation } from "../types";
import { ReactNode } from "react";

interface StationProps {
  onSelect: (station: IStation) => ReactNode;
  onToggleFavorite: (station: IStation) => void;
  isFavorite?: boolean;
  station: IStation;
}

export default function Station({ isFavorite, onSelect, onToggleFavorite, station }: StationProps) {
  return (
    <List.Item
      accessories={[
        ...(isFavorite ? [{ icon: { source: Icon.Heart, tintColor: Color.Red }, tooltip: "Favorite station" }] : []),
        { icon: Icon.ArrowRight },
      ]}
      icon={Icon.Geopin}
      actions={
        <ActionPanel>
          <Action.Push title="Select" target={onSelect(station)} icon={Icon.ArrowRight} />
          <Action
            title={isFavorite ? "Remove from Favorites" : "Add to Favorites"}
            icon={{
              source: isFavorite ? Icon.HeartDisabled : Icon.Heart,
              tintColor: Color.Red,
            }}
            shortcut={{ modifiers: isFavorite ? ["cmd", "shift"] : ["cmd"], key: "s" }}
            onAction={() => {
              onToggleFavorite(station);
            }}
          />
        </ActionPanel>
      }
      subtitle={station.Location}
      title={station.Name}
    />
  );
}
```

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
I added a few corrections, feel free to implement 😊

File: extensions/stockholm-public-transport/src/components/station.tsx (hunk 3 above)
Comment (albingroen) [reply extensions/stockholm-public-transport/src/components/station.tsx]: ✓

File: extensions/stockholm-public-transport/src/lib/stations.ts (hunk 2 above)
Comment (albingroen) [reply extensions/stockholm-public-transport/src/lib/stations.ts]: ✓

File: extensions/stockholm-public-transport/src/components/stations.tsx (hunk 1 above)
Comment (albingroen) [reply extensions/stockholm-public-transport/src/components/stations.tsx]: ✓

**Review by pernielsentikaer — APPROVED** [review]
Hi 👋 

Looks good to me, approved 🔥

**Pull request merged**